File: src/connectors/urlscan.js
import { BaseConnector } from './base.js';
import https from 'https';

export class URLScanConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'URLScan';
        this.apiKey = process.env.URLSCAN_API_KEY;
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        try {
            const entities = [];
            const query = inputType === 'domain' ? `domain:${inputValue}` : `page.url:"${inputValue}"`;
            const results = await this._apiGet(`/api/v1/search/?q=${encodeURIComponent(query)}&size=20`);

            if (results?.results) {
                const seenIPs = new Set();
                const seenDomains = new Set();
                const technologies = new Set();

                for (const scan of results.results) {
                    const page = scan.page || {};

                    if (page.ip && !seenIPs.has(page.ip)) {
                        seenIPs.add(page.ip);
                        entities.push({
                            type: 'ip',
                            data: {
                                address: page.ip,
                                asn: page.asnname,
                                country: page.country,
                                server: page.server,
                                source: 'urlscan'
                            },
                            confidence: 0.8,
                            asKnown: { type: 'ip', value: page.ip }
                        });
                    }

                    if (page.domain && !seenDomains.has(page.domain) && page.domain !== inputValue) {
                        seenDomains.add(page.domain);
                        entities.push({
                            type: 'domain',
                            data: { name: page.domain, source: 'urlscan' },
                            confidence: 0.7,
                            asKnown: { type: 'domain', value: page.domain }
                        });
                    }

                    if (page.server) technologies.add(page.server);
                }

                if (technologies.size > 0) {
                    entities.push({
                        type: 'technology',
                        data: {
                            url: inputType === 'domain' ? `https://${inputValue}` : inputValue,
                            technologies: [...technologies],
                            source: 'urlscan'
                        },
                        confidence: 0.7
                    });
                }

                if (inputType === 'domain') {
                    const latest = results.results[0];
                    if (latest?.task?.uuid) {
                        const detail = await this._apiGet(`/api/v1/result/${latest.task.uuid}/`);
                        if (detail?.lists) {
                            for (const domain of (detail.lists.domains || []).slice(0, 20)) {
                                if (!seenDomains.has(domain) && domain !== inputValue) {
                                    seenDomains.add(domain);
                                    entities.push({
                                        type: 'domain',
                                        data: { name: domain, source: 'urlscan_resources' },
                                        confidence: 0.5
                                    });
                                }
                            }

                            for (const ip of (detail.lists.ips || []).slice(0, 20)) {
                                if (!seenIPs.has(ip)) {
                                    seenIPs.add(ip);
                                    entities.push({
                                        type: 'ip',
                                        data: { address: ip, source: 'urlscan_resources' },
                                        confidence: 0.5,
                                        asKnown: { type: 'ip', value: ip }
                                    });
                                }
                            }

                            if (detail.lists.certificates?.length) {
                                for (const cert of detail.lists.certificates.slice(0, 10)) {
                                    entities.push({
                                        type: 'certificate',
                                        data: {
                                            subject: cert.subjectName,
                                            issuer: cert.issuer,
                                            valid_from: cert.validFrom ? new Date(cert.validFrom * 1000).toISOString() : null,
                                            valid_to: cert.validTo ? new Date(cert.validTo * 1000).toISOString() : null,
                                            source: 'urlscan'
                                        },
                                        confidence: 0.8
                                    });
                                }
                            }
                        }
                    }
                }
            }

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
                if (entity.asKnown) {
                    this.state.addKnown(investigationId, entity.asKnown.type, entity.asKnown.value, this.name, entity.confidence);
                }
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, null, null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    _apiGet(path) {
        return new Promise((resolve, reject) => {
            const headers = { 'Accept': 'application/json' };
            if (this.apiKey) headers['API-Key'] = this.apiKey;

            const req = https.get({
                hostname: 'urlscan.io',
                path,
                headers,
                timeout: 15000
            }, res => {
                if (res.statusCode === 404) { res.resume(); return resolve(null); }
                if (res.statusCode === 429) { res.resume(); return reject(new Error('URLScan rate limited')); }
                if (res.statusCode >= 400) { res.resume(); return reject(new Error(`URLScan HTTP ${res.statusCode}`)); }
                let data = '';
                res.on('data', chunk => { data += chunk; });
                res.on('end', () => {
                    try { resolve(JSON.parse(data)); } catch { resolve(null); }
                });
            });
            req.on('error', reject);
            req.on('timeout', () => { req.destroy(); reject(new Error('URLScan timeout')); });
        });
    }

    async healthCheck() {
        return { ok: true, version: 'v1', hasApiKey: !!this.apiKey };
    }
}

File: cyclops/connectors/urlscan.py
import os
import aiohttp
from urllib.parse import quote
from cyclops.connectors.base import BaseConnector


class URLScanConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "URLScan"
        self.api_key = os.environ.get("URLSCAN_API_KEY")

    async def run(self, investigation_id, phase_id, input_type, input_value):
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})
        try:
            entities = []
            headers = {"Accept": "application/json"}
            if self.api_key:
                headers["API-Key"] = self.api_key
            query = f"domain:{input_value}" if input_type == "domain" else f'page.url:"{input_value}"'
            async with aiohttp.ClientSession(headers=headers) as session:
                async with session.get(f"https://urlscan.io/api/v1/search/?q={quote(query)}&size=20", timeout=aiohttp.ClientTimeout(total=15)) as resp:
                    if resp.status >= 400:
                        raise Exception(f"URLScan HTTP {resp.status}")
                    data = await resp.json()
            seen_ips, seen_domains = set(), set()
            for scan in (data or {}).get("results", []):
                page = scan.get("page", {})
                if page.get("ip") and page["ip"] not in seen_ips:
                    seen_ips.add(page["ip"])
                    entities.append({"type": "ip", "data": {"address": page["ip"], "asn": page.get("asnname"), "country": page.get("country"), "server": page.get("server"), "source": "urlscan"}, "confidence": 0.8, "as_known": {"type": "ip", "value": page["ip"]}})
                if page.get("domain") and page["domain"] != input_value and page["domain"] not in seen_domains:
                    seen_domains.add(page["domain"])
                    entities.append({"type": "domain", "data": {"name": page["domain"], "source": "urlscan"}, "confidence": 0.7})
            new_count = 0
            for entity in entities:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)
                if "as_known" in entity:
                    self.state.add_known(investigation_id, entity["as_known"]["type"], entity["as_known"]["value"], self.name)
            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": entities, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    async def health_check(self):
        return {"ok": True, "hasApiKey": bool(self.api_key)}
